src/DateRangePicker.ts

```diff
@@ -1,26 +1,423 @@
+  showTooltip(e: PointerEvent): void {
+    const target = e.currentTarget as SVGRectElement;
+    const x = target.x.baseVal.value + this.sliderWidth / 2;
+    const data = target.dataset;
+    const itemsText = `item${data.numItems !== '1' ? 's' : ''}`;
+    const tooltipOffset =
+      x + (this._binWidth - this.sliderWidth - this.tooltipWidth) / 2;
+    this.tooltip.style.left = `${tooltipOffset}px`;
+    this.tooltip.innerHTML = `${data.numItems} ${itemsText}<br>${data.binStart} - ${data.binEnd}`;
+    this.tooltip.style.display = 'block';
+  }
+
+  hideTooltip(): void {
+    this.tooltip.style.display = 'none';
+    this.tooltip.innerHTML = '';
+  }
+
+  @query('#histogram') histogram!: HTMLDivElement;
+
+  drag(e: PointerEvent): void {
+    // prevent selecting text or other ranges while dragging, especially in Safari
+    e.preventDefault();
+
+    this.currentSlider = e.currentTarget as SVGRectElement;
+    const sliderX =
+      (this.currentSlider.id as SliderIds) === 'slider-min'
+        ? this._leftSliderX
+        : this._rightSliderX;
+    this._dragOffset = e.offsetX - sliderX;
+    this.currentSlider.classList.add('dragging');
+    this.histogram.addEventListener('pointermove', this.move);
+
+    // ensure that we catch the pointer up event, no matter where the pointer is
+    // when it's released. for more info see
+    // https://developers.google.com/web/updates/2016/10/pointer-events#pointer_capture
+    if (e.pointerId) {
+      this.histogram.setPointerCapture(e.pointerId);
+    }
   }
 
-  render() {
+  drop(): void {
+    this.histogram.removeEventListener('pointermove', this.move);
+    if (this.currentSlider) {
+      this.currentSlider.classList.remove('dragging');
+    }
+  }
+
+  // use an arrow function so that `this` is bound to the DateRangePicker object
+  // and not the event target. for more info see
+  // https://lit-element.polymer-project.org/guide/events#using-this-in-event-listeners
+  move = (e: PointerEvent): void => {
+    const newX = e.offsetX - this._dragOffset;
+    const slider = this.currentSlider as SVGRectElement;
+    return (slider.id as SliderIds) === 'slider-min'
+      ? this.setLeftSlider(newX)
+      : this.setRightSlider(newX);
+  };
+
+  setLeftSlider(newX: number): void {
+    const toSet = Math.min(newX, this._rightSliderX);
+    this._leftSliderX = Math.max(toSet, this.sliderWidth);
+  }
+
+  setRightSlider(newX: number): void {
+    const toSet = Math.max(newX, this._leftSliderX);
+    this._rightSliderX = Math.min(toSet, this.width - this.sliderWidth);
+  }
+
+  translatePositionToDate(x: number): string {
+    const milliseconds =
+      ((x - this.sliderWidth) * this.dateRange()) / this._histWidth;
+    const date = dayjs(this._minDate + milliseconds);
+    return date.format(this.dateFormat);
+  }
+
+  translateDateToPosition(date: string): number {
+    const milliseconds: number | undefined = dayjs(date).valueOf();
+    if (!milliseconds) {
+      return DATE_PARSING_ERROR;
```

Comment:
Instead of passing back a magic value, you could just make the return value optional and return nothing if this fails.